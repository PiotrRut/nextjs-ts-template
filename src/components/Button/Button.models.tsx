import { ButtonHTMLAttributes } from 'react';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  name: HTMLButtonElement['name'];
  /** Indicates whether the button is in loading state where it shows a spinner */
  disabled?: boolean;
  /** When true, the button will occupy 100% width of its parent */
  fullWidth?: boolean;
}
