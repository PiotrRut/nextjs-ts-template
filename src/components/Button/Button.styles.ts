/* eslint-disable indent */
import { buttonReset, centredFlex } from '@theme/shortcuts';
import { sizes } from '@theme/sizes';
import styled, { css } from 'styled-components';

import { ButtonProps } from './Button.models';

export const ButtonContainer = styled.button<ButtonProps>`
  ${buttonReset}
  ${centredFlex}
  width: ${(p) => (p.fullWidth ? '100%' : css`calc(100vw - ${sizes[40]})`)};
  height: ${sizes[50]};
  background-color: ${(p) => (p.disabled ? 'gray' : 'green')};
  transition-property: background-color;
  transition-duration: 0.3s;
  transition-timing-function: ease-in-out;
  pointer-events: ${(p) => (p.disabled ? 'none' : 'auto')};
  color: white;
  &:hover {
    background-color: ${(p) => (p.disabled ? 'gray' : 'blue')};
    cursor: ${(p) => (p.disabled ? 'auto' : 'pointer')};
  }
  @media (min-width: 480px) {
    width: ${(p) => (p.fullWidth ? '100%' : sizes[350])};
  }
`;
