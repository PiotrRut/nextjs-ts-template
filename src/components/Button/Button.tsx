import React, { forwardRef } from 'react';

import { ButtonProps } from './Button.models';
import { ButtonContainer } from './Button.styles';

const Button = forwardRef<HTMLButtonElement, ButtonProps>((props, ref) => {
  const {
    disabled = false,
    children,
    fullWidth = false,
    name,
    ...rest
  } = props;

  return (
    <ButtonContainer
      {...{ disabled, fullWidth, ref, ...rest }}
      data-testid={name}
      name={name}
    >
      {children}
    </ButtonContainer>
  );
});

export default Button;
