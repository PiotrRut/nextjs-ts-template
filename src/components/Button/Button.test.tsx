import { render } from '@testing-library/react';
import React from 'react';

import Button from './Button';

test('Button Test', () => {
  const { getByTestId, container } = render(<Button name="test">Hello</Button>);

  const buttonNode = getByTestId('test');

  expect(buttonNode.textContent).toBe('Hello');

  expect(container).toMatchSnapshot();
});
