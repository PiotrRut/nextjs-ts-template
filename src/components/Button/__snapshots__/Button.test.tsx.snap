// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Test 1`] = `
<div>
  <button
    class="sc-bdvvtL hoPvan"
    data-testid="test"
    name="test"
  >
    Hello
  </button>
</div>
`;
